subsys/power/include/pm_policy.h

```diff
@@ -0,0 +1,182 @@
+#define _PM_POLICY_H_
+
+#include <string.h>
+#include <sys/util.h>
+#include <power_state.h>
+#include <toolchain/common.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Power management policy
+ * @defgroup Pm_policy Power management policy
+ * @{
+ * @}
+ */
+struct pm_policy;
+
+/**
+ * @brief Power management policy API.
+ */
+struct pm_policy_api {
+	/* initialize the power management policy */
+	void (*init)(void);
+
+	/* get next system power state based on current system
+	 * status and the power management policy where this
+	 * api resides.
+	 */
+	pm_state_t (*next_state)(struct pm_policy *policy);
+
+	/* set constraints for the supported power states so
+	 * that others can impact the pm_policy when make decision.
+	 */ 
+	int (*set_constraint)(struct pm_policy *policy, pm_state_t states);
+
+	/* release the constraints set by set_constraint for
+	 * the supported power states.
+	 * should be paired with set_constraint.
+	 */
+	int (*release_constraint)(struct pm_policy *policy, pm_state_t states);
+};
+
+/**
+ * @brief Power management policy structure.
+ */
+struct pm_policy {
+	pm_state_t supported_states;
+	struct pm_policy_api *policy_api;
+};
+
+/**
+ * @brief Create pm_policy instance.
+ *
+ * @param _name Instance name.
+ * @param _supported_states Supported states.
+ * @param _api Power management policy API.
+ */
+#define PM_POLICY_DEFINE(_name, supported_states, _api) \
+	static const Z_STRUCT_SECTION_ITERABLE(pm_policy, _name) = \
+	{ \
+		.name = STRINGIFY(_name), \
+		.supported_states = (_supported_states & PM_STATE_BIT_MASK), \
+		.policy_api = &_api \
+	}
+
+/**
+ * @brief Initialize power management policy.
+ *
+ * @param policy Pointer to pm_policy instance.
+ */
+static inline void pm_policy_init(struct pm_policy *policy)
+{
+	if (policy && policy->api) {
+		policy->api->init();
+	}
+}
+
+/**
+ * @brief Generate system power state based on the given policy.
+ *
+ * @param policy Pointer to pm_policy instance.
+ *
+ * @return Power state of system decided by the policy or
+ * 	   PM_STATE_RUNTIME_IDLE if not completed structure.
+ */
+static inline pm_state_t pm_policy_next_state(struct pm_policy *policy)
+{
+	if (policy && policy->api) {
+		return policy->api->next_state(policy);
+	}
+
+	return PM_STATE_RUNTIME_IDLE;
+}
+
+/**
+ * @brief Get supported power states based on the given policy.
+ *
+ * @param policy Pointer to pm_policy instance.
+ *
+ * @return Supported power states of the given policy or if not completed
+ * 	   structure return PM_STATE_RUNTIME_IDLE | PM_STATE_RUNTIME_ACTIVE.
+ */
+static inline pm_state_t pm_policy_get_supported_state(struct pm_policy *policy)
+{
+	if (policy) {
+		return policy->supported_states & PM_STATE_BIT_MASK;
+	}
+
+	return PM_STATE_RUNTIME_ACTIVE | PM_STATE_RUNTIME_IDLE;
+}
+
+/**
+ * @brief Set constraint for the power states of the given policy.
+ *
+ * @param policy Pointer to pm_policy instance.
+ * @param states Power states on which set constraint.
```

Comment:
What is a constraint?

If a state is selected in the bitset is it a state that can be entered, or a state that cannot be entered?

Maybe this should be:

> @param states A bit set identifying PM states that the power management policy is not permitted to select.

Reply:
@pabigot a state that cannot be entered. Will update the comments much clear. Thanks!